### PR TITLE
Updates for Offline Deployment and General Bug/QoL Fixes

### PR DESCRIPTION
**Pull request by tylera-nvidia**
- Adding all Thirdparty packages to CPM for ease of deployment
- Updating CPM and adding as git subtree of project
- Documentation on all of the above _(in progress)_
- Fixes for NVTX Macros and CMake flags 
- Update includes for einsum
- Fix for size checks on matvec
- Documentation on linkage/usage

### CODE REVIEW

File: test/include/test_types.h [hunk 1]
```diff
@@ -78,9 +78,8 @@ typedef Types<matx::matxFp16, matx::matxBf16, bool, uint32_t, int32_t, uint64_t,
               cuda::std::complex<double>, matx::matxFp16Complex,
               matx::matxBf16Complex>
     MatXAllTypes;
-typedef Types<matx::matxFp16, matx::matxBf16, float, double,
-              cuda::std::complex<float>, cuda::std::complex<double>,
-              matx::matxFp16Complex, matx::matxBf16Complex>
+typedef Types<float, double,
```
Comment (cliffburdick) [inline test/include/test_types.h:81]: What's the reason behind this change? This seems like it would reduce test coverage. Right below it is a `MatXFloatNonHalfTypes` that has the exact same types.

File: docs_input/build.rst [hunk 2]
```diff
@@ -15,17 +16,30 @@ from the internet. Alternatively, the option ``CPM_USE_LOCAL_PACKAGES`` can be u
...
 
-Core Requirements
+System Requirements
 -----------------
 MatX requires **CUDA 11.4** or higher, and **g++ 9.3** or higher for the host compiler. Clang may work as well, but it's currently 
 untested. Other requirements for optional components are listed below.
 
 .. warning:: Using MatX with an unsupported compiler may result in compiler and/or runtime errors.
 
+Required Thirdparty Dependencies
```
Comment (cliffburdick) [inline docs_input/build.rst:27]: Third-party

File: docs_input/build.rst [hunk 3]
```diff
@@ -15,17 +16,30 @@ from the internet. Alternatively, the option ``CPM_USE_LOCAL_PACKAGES`` can be u
...
+
+Optional Thirdparty Dependencies
+-----------------
+.. _GoogleTest: https://github.com/google/googletest
+.. _pybind11: https://github.com/pybind/pybind11
+nvbench
+cutensor
+cutensornet
+fmt
+nloghmann
```
Comment (cliffburdick) [inline docs_input/build.rst]: nlohmann::json

File: docs_input/build.rst [hunk 4]
```diff
@@ -120,3 +134,56 @@ to browse the documentation. Note that the most recent version of the documentat
 
 https://nvidia.github.io/MatX/
 
+MatX in Offline Environments
```
Comment (cliffburdick) [inline docs_input/build.rst:137]: I would put this section towards the bottom since it's likely not as useful to the public.

File: docs_input/build.rst [hunk 5]
```diff
@@ -120,3 +134,56 @@ to browse the documentation. Note that the most recent version of the documentat
...
+
+    
+- Build your MatX project per your standard process, CPM will automatically use the cache
+
+
+
+MatX Library Linking
+=============
+MatX defaults to Hidden Visibility due to compile requirements from pybind (https://pybind11.readthedocs.io/en/stable/faq.html#someclass-declared-with-greater-visibility-than-the-type-of-its-field-someclass-member-wattributes). 
+Hidden Visibility hides symbols from the C++ linker, and will prevent a user from accessing functions from other compile units. If inheriting the MatX Build system, this will also prevent user-space symbols from being enabled, which may be a problem for multi-library or resource projects intended
```
Comment (cliffburdick) [inline docs_input/build.rst:188]: compile units should be translation units

**Review by cliffburdick — APPROVED** [review]
Looks good!

**Comment by tylera-nvidia** [comment]
[----------] Global test environment tear-down
[==========] 1255 tests from 349 test suites ran. (673523 ms total)
[  PASSED  ] 1255 tests.

  YOU HAVE 20 DISABLED TESTS

docker@ipp1-1946:/home/tallen/scratch/MatX/build$ git status
On branch offline
Your branch is up to date with 'origin/offline'.

nothing to commit, working tree clean
docker@ipp1-1946:/home/tallen/scratch/MatX/build$ git rev-parse HEAD
020e2341337d01898c0e52146cdbf163b370c11c
docker@ipp1-1946:/home/tallen/scratch/MatX/build$
